server/src/main/java/org/opensearch/gateway/BaseShardCache.java

```diff
@@ -0,0 +1,291 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.gateway;
+
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.ExceptionsHelper;
+import org.opensearch.OpenSearchTimeoutException;
+import org.opensearch.action.FailedNodeException;
+import org.opensearch.action.support.nodes.BaseNodeResponse;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.cluster.node.DiscoveryNodes;
+import org.opensearch.core.concurrency.OpenSearchRejectedExecutionException;
+import org.opensearch.transport.ReceiveTimeoutTransportException;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Common functionalities of a cache for storing shard metadata. Cache maintains node level responses.
+ * Setting up the cache is required from implementation class.
+ *
+ * @param <K> Response type of transport action which has the data to be stored in the cache.
+ */
+public abstract class BaseShardCache<K extends BaseNodeResponse> implements NodeCache<K> {
```

Comment:
Why does BaseShardCache implement NodeCache if it is not actually implementing any of the methods in the interface? 

Assume that, we move methods in NodeCache to BaseShardCache as abstract methods, what do we lose?

Reply:
>Assume that, we move methods in NodeCache to BaseShardCache as abstract methods, what do we lose? 

Actually it was that way only in last commit 🙂  https://github.com/opensearch-project/OpenSearch/pull/12441/files/11e145d1c4f3557466f9e0ba425f323f29748a44 

I wanted to separate the interface and keep common functionalities in BaseShardCache abstract class. 
The implementation of interface method is done by `ShardCache` class, and common code written in `BaseShardCache`(abstract one) uses those methods of interface like `getData, putData or getCache` so it's implementing the interface. 

Other implementation of this `NodeCache` **interface** will be **`ShardBatchCache`** which is written in https://github.com/opensearch-project/OpenSearch/pull/12504/files#diff-9f14f4320d39587e9cb6d8a76bed25ed1a2ee8464246700ccb3a3777b4ccb516

@sachinpkale - what do you suggest ? Should I revert the code to keep all those method in abstract class only how it was before or keep it in the interface so the contract is clearly visible for users.

Reply:
I'm planning to merge back the `NodeCache` methods into **BaseShardCache** how it was before. And will delete NodeCache file.